add regularizer

hkirvesl opened the pull request:
**Reference issues/PRs**

This Pull Request adds the regularization logic, fixing issue #131

**Types of changes**

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

**Description**

This PR introduces the regularization logic, which enables the training loss to depend on more general parameters
other than the input data and response.

To this end, a regularizer class was implemented, together with modifications to the trainer class. The hpo was also updated to accommodate regularizer parameters.

The framework is quite flexible, and comes with pre-existing implementation of the so called p-norm regularizer.

The main features of this PR are presented in 2 jupyter notebooks, one for general introduction to the topic and another one for implementing custom regularizers.

The version of scikit-learn was set to 1.1.1, because newer versions don't have the delegate_has_method decorator that breaks a functionality it gtda that this PR depends on. 

In addition, Pillow was downgraded to 9.5.0 as certain function unrelated to this PR were deprecated in 10.0.0 


**Checklist**

- [X] I have read the [guidelines for contributing](https://giotto-ai.github.io/gtda-docs/latest/contributing/#guidelines).
- [X] My code follows the code style of this project. I used `flake8` to check my Python changes.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed. I used `pytest` to check this on Python tests.

raphaelreinauer commented on gdeep/trainer/regularizer.py:
```diff
@@ -0,0 +1,432 @@
+        self.regularizer_arguments["dummy_loader"] = dummy_loader
+        return True
+
+    def update_params(self, regularizer_arguments):
```
I would introduce a dataclass here. This makes the code much easier to use and is less error prone.

hkirvesl replied:
done

raphaelreinauer commented on gdeep/trainer/trainer.py:
```diff
@@ -304,7 +320,23 @@ def closure() -> Tensor:
-            loss = self.loss_fn(pred, y)
+            if self.regularize == True:
+                loss = self.loss_fn(pred, y)
+                self.regularizer.regularizer_arguments["model"] = copy.deepcopy(
```
Why is the deepcopy necessary here?

AnthoJack replied:
This is purely esthetic but don't use `==True`. `self.regularize` is already a boolean

hkirvesl replied:
right, it's not

raphaelreinauer commented on gdeep/trainer/regularizer.py:
```diff
@@ -0,0 +1,432 @@
+
+
+class RegularizedTrainer(Trainer):
+    def __init__(self, regularizer, reg_params={}, *args, **kwargs):
```
Add type hint to the parameter `regularizer`.

hkirvesl replied:
done, (class removed)

raphaelreinauer reviewed (commented):
I added some comments about the inheritance approach for `RegularizedTrainer`.

raphaelreinauer commented on gdeep/search/hpo_config.py:
```diff
@@ -11,7 +11,7 @@ class HPOConfig:
     Args:
         optimizers:
-            list of torch optimizers classes, not instances
+            list of torch optimizers classes, not isntances
```
looks like a typo

hkirvesl replied:
done

raphaelreinauer commented on gdeep/search/hpo_config.py:
```diff
@@ -39,10 +39,10 @@ class HPOConfig:
             tensorboard histograms
         writer_tag:
-            tag to prepend to the output
+            tag to prepend to the ouput
```
here as well

hkirvesl replied:
done

raphaelreinauer commented on gdeep/search/hpo_config.py:
```diff
@@ -23,7 +23,7 @@ class HPOConfig:
             dictionary of model parameters
         lr_scheduler:
-            torch learning rate scheduler class
+            torch learning rate schduler class
```
here as well

hkirvesl replied:
done

raphaelreinauer commented on gdeep/search/hpo_config.py:
```diff
@@ -39,10 +39,10 @@ class HPOConfig:
             training or not
         store_grad_layer_hist:
-            flag to store the gradients of the layers in the
+            flag to store the gradents of the layers in the
```
typo

hkirvesl replied:
done

raphaelreinauer commented on gdeep/search/hpo.py:
```diff
@@ -615,7 +638,7 @@ def _store_to_list_each_step(
                 the value of the loss for the current trial
             accuracy (float, default -1):
-                the value of the accuracy for the current trial
+                the value of the accuracy for te current trial
```
typo

hkirvesl replied:
done

raphaelreinauer commented on gdeep/trainer/trainer.py:
```diff
@@ -304,7 +320,23 @@ def closure() -> Tensor:
+                self.regularizer.regularizer_arguments["model"] = copy.deepcopy(
+                    self.model
+                )
+                self.regularizer.regularizer_arguments[
```
What's the type of self.regularizer. The default value is "Regularizer"`.

matteocao replied:
I believe the type is `Regularizer`, as per line 39 and 148.

hkirvesl replied:
correct

raphaelreinauer commented on gdeep/trainer/regularizer.py:
```diff
@@ -0,0 +1,432 @@
+        )
+        cpoints = np.concatenate(
+            [cpoints1.detach().numpy(), cpoints2.detach().numpy()], 0
+        )
```
What's the difference between `cpoints`, `cpoints1` and `cpoints2`?

hkirvesl replied:
The cpoints1 and cpoints2 are (select) critical points of f and -f, respectively. Cpoints is then both of them

raphaelreinauer replied:
Could you please adjust the names to reflect that.

hkirvesl replied:
done

matteocao commented on gdeep/trainer/regularizer.py:
```diff
@@ -0,0 +1,432 @@
+    A helper file for topological regularizer
+    Computes the local homology, used for the linear surrogate function
+    Args:
+        X: The point cloud on which the surrogate function is evaluated
```
please put the description one line below, so that it appears nicely on the sphinx html.

like so:
```
Args:
    X:
        The point cloud ....
    param2:
        The description of ....
```

hkirvesl replied:
done

matteocao commented on gdeep/trainer/regularizer.py:
```diff
@@ -0,0 +1,432 @@
+        dim: the homology dimension based on which the neighbors are computed
+             In the case of the topological regularizer, this is 1.
+    Returns
+    -------
```
no need for this line `-------` . please remove

hkirvesl replied:
done

matteocao commented on gdeep/trainer/regularizer.py:
```diff
@@ -0,0 +1,432 @@
+    results: np.ndarray, connections: np.ndarray, expansion: int = 6
+) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    A helper file for topological regularizer
```
is the user supposed to use all these helper functions?
If they are useful only in the topological regularizer, then please start their name with `_` , as to indicate the user that they shall not use these functions directly. For example `def _get_persistent_pairs(...)`

hkirvesl replied:
done

matteocao commented on gdeep/trainer/regularizer.py:
```diff
@@ -0,0 +1,432 @@
+    returns the persistent pairs of the function
+    associated to filtering according to the numerical values
+    Args:
+        results: an n-vector of class 1 probabilities
```
same cimment as above on formatting. please skip a line and add 4 spaces

hkirvesl replied:
done

matteocao commented on gdeep/trainer/regularizer.py:
```diff
@@ -0,0 +1,432 @@
+        among the entries of results
+        expansion: The size of the clique expansion when computing homology
+    Returns
+    -------
```
as above, please remove

hkirvesl replied:
done

matteocao commented on gdeep/trainer/regularizer.py:
```diff
@@ -0,0 +1,432 @@
+    persistences = np.zeros([len(tmp), 2])
+    pers_indices = np.zeros([len(tmp), 2])
+    odd_one_out = 0
+    for i in range(len(tmp)):
```
can you please add a couple of inline comments to describe what your algorithm is doing here?

hkirvesl replied:
done

matteocao commented on gdeep/trainer/regularizer.py:
```diff
@@ -0,0 +1,432 @@
+    return (out_pers, persts)
+
+
+def compute_critical_points(
```
if this helper function is not meant to be used directly by the user, then please start its name with `_`

hkirvesl replied:
done

matteocao commented on gdeep/trainer/regularizer.py:
```diff
@@ -0,0 +1,432 @@
+    The filtering directions is specified by ind (0: Positive, 1 negative)
+    A helper file for topological regularizer
+    Args:
+        model: torch model
```
same as above on foramtting

hkirvesl replied:
done

matteocao commented on gdeep/trainer/regularizer.py:
```diff
@@ -0,0 +1,432 @@
+            0: positive filtration, 1: negative filtration
+    Returns
+    -------
+        coordinates: The coordinates of the critical points that
```
formatting, as above

hkirvesl replied:
done

matteocao commented on gdeep/trainer/regularizer.py:
```diff
@@ -0,0 +1,432 @@
+    def __init__(self, **kwargs):
+        self.regularizer_arguments = kwargs
+
+    """
```
this description is for the class, hence please put it before the `__init__()`, as it is the docstring for the class itself

hkirvesl replied:
done

matteocao commented on gdeep/trainer/regularizer.py:
```diff
@@ -0,0 +1,432 @@
+            self.regularizer_arguments["regularization_dataset"]
+        )
+        connections = local_homology_preprocess(
+            self.regularizer_arguments["regularization_dataset"],
```
can you please define a `@dataclass` for the regularizer parameters for the topological regularizer where you put all these parameters in? the dataclass does not need to have any method, just store all the needed parameters with type.

hkirvesl replied:
done, it ended up having `__post_init__` to achieve what was previously known as `preprocess`

matteocao commented on gdeep/trainer/trainer.py:
```diff
@@ -132,6 +145,8 @@ def forward(self, x):
         Callable[[int, Optimizer, ModelExtractor, Optional[SummaryWriter]], Any]
     ] = None
+    regularizer: "Regularizer"
+    lamda: float
```
I think this lambda has to have a default of, say, `1.`

hkirvesl replied:
Done. In the end this lamda was moved to be a property of a regularizer, not trainer. This makes the regularizer parameters more similar (you pass them all to the regularizer). Also, as it is not absolutely required for all regularizers, this avoids the tricks previously present in defining e.g. elastic net that uses several lambdas

matteocao commented on gdeep/search/hpo.py:
```diff
@@ -306,14 +311,30 @@ def _objective(self, trial: BaseTrial, config: HPOConfig):
-            self.training_metric,
-            self.k_fold_class,
-        )
+        if self.regularize == True:
```
@hkirvesl please add a unit test for the hpo (a very simple one) that tests this pathway

raphaelreinauer replied:
Having to add `RegulariedTrainer` in parts of the code that does not have anything to do with regularization looks very fishy.

hkirvesl replied:
done

matteocao commented on examples/deploying_custom_regularizers.ipynb:
```diff
@@ -0,0 +1,626 @@
+    "\n",
+    "The diagram below gives an overview of these steps:\n",
+    "\n",
+    "![img](./images/reg_sw_diagram.png)\n"
```
I think you forgot to add the picture in you PR, as I do not see it

hkirvesl replied:
Right. In the end the implementation got much simplified, so I think there is no longer need for the diagram. The plain text in the notebook is changed accordingly.

raphaelreinauer commented on gdeep/trainer/regularizer.py:
```diff
@@ -0,0 +1,432 @@
+    return dummy_loader
+
+
+def unique_list(a_list: list) -> list:
```
This should be simplified to `list(set(a_list))`

hkirvesl replied:
You mean the function? That would work if the elements of a_list were hashable. In this case they are not.

raphaelreinauer commented on gdeep/trainer/regularizer.py:
```diff
@@ -0,0 +1,432 @@
+    return uniquelist
+
+
+def evaluate_model_on_grid(
```
I don't understand the grid in the function's name.

hkirvesl replied:
Right, the input needs not be a grid. However, functionally it does exactly that. The namesake is what is described in https://arxiv.org/pdf/1806.10714.pdf page 4 under algorithm (first and last paragraphs)

matteocao reviewed (changes requested):
@hkirvesl thank you very much for your contribution!
There are a few changes needed, not major ones, imho. You will find all the comments in the code.

However, please answer carefully the question Raphael asked about the "overwriting of the loss_fn" instead of building the new RegularizedTrainer. This is very valid point and maybe adding that class is too much, as we can simply add two arguments (i.e. `regularizer, reg_params={}`) to the trainer class and default them to None.

raphaelreinauer commented on gdeep/trainer/regularizer.py:
```diff
@@ -0,0 +1,432 @@
+            value = results[inds1[index]]
+            if value > persistences[i, 0]:
+                persistences[i, 0] = value
+                pers_indices[i, 0] = inds1[index]
```
naming is not consistent `persistences` and `pers` and `persts` also I think full names would make thing much easier to understand.

hkirvesl replied:
done

raphaelreinauer commented on gdeep/trainer/regularizer.py:
```diff
@@ -0,0 +1,432 @@
+    cuts = (tmp[:, 0] * tmp[:, 1]) < 0
+
+    goodinds = np.argmin(abs(tmp[cuts, :]), 1)
+    tmp_inds = persts[cuts, :]
```
also here the naming is not consistent `goodinds` without `_ ` and `tmp_inds` with `_`.

raphaelreinauer replied:
Also `good` is not very descriptive.

hkirvesl replied:
done

raphaelreinauer commented on gdeep/trainer/regularizer.py:
```diff
@@ -0,0 +1,432 @@
+    A topological regularizer for a 2-class classifier
+    """
+
+    def preprocess(self, **kwargs):
```
I think it's better to put all of this in the `__init__` and have proper function parameters.

hkirvesl replied:
done

raphaelreinauer commented on gdeep/search/hpo.py:
```diff
@@ -216,6 +217,7 @@ def __init__(
         self.scalars_dict: Dict[str, Any] = dict()
         # can be changed by changing this attribute
         self.store_pickle: bool = False
+        self.regularize = False
```
How would I set regularize to `True` from the outside?

hkirvesl replied:
done

raphaelreinauer commented on gdeep/trainer/tests/test_regularizer.py:
```diff
@@ -0,0 +1,117 @@
+
+
+# model
+class Model1(nn.Module):
```
Give it a more descriptive name like `SmallFullyConnectedNeuralNetwork`

hkirvesl replied:
this is a verbatim copy of the test_trainer.py, so i left it like that

raphaelreinauer replied:
I think it would make sense to add the tests to `test_trainer.py`, then you don't have to copy code.

raphaelreinauer commented on gdeep/trainer/__init__.py:
```diff
@@ -1,6 +1,9 @@
 from .trainer import Trainer
 from .trainer_config import TrainerConfig
-
+from .regularizer import Regularizer
```
Why are these imports needed?

hkirvesl replied:
forgot to clean... done

raphaelreinauer commented on gdeep/trainer/regularizer.py:
```diff
@@ -0,0 +1,432 @@
+        Computes the actual regression penalty.
+        This must be an auto-differentiable function of its inputs
+        """
+        return torch.sum(forward_passed_arguments) ** 2
```
This method is abstract, so replace implementation by `pass`

hkirvesl replied:
done

raphaelreinauer commented on gdeep/trainer/regularizer.py:
```diff
@@ -0,0 +1,432 @@
+    out_pers = np.delete(persistences, odd_one_out, axis=0)
+    out_pers_ind = np.delete(pers_indices, odd_one_out, axis=0)
+    persts = out_pers_ind.astype(int)
+    return (out_pers, persts)
```
what's the difference between `out_pers` and `persts`?

hkirvesl replied:
added explanations

hkirvesl commented:
Thanks Matteo, Raphael and AnthoJack for taking the time to comment the code. I think the code quality improved by a lot.

I've changed the logic so that RegularizedTrainer was removed, and everything that has anything to do with regularization is in the Regularizer class that can be passed to the Trainer.

For the comment on overwriting self.loss_fn function in trainer, I think this may be unwise, if you intend to use this loss for validation, and it serves as some sort of metric of goodness of fit. An apples-to-apples comparison would be good there, independent of say lamda. The logic suggested here is add the penalty term in training if a regularizer is given. I believe this is a good compromise

raphaelreinauer commented:
> self.loss_fn function in trainer, I think this may be unwise, if you intend to use this loss for validation

The best would be to have a train_loss and a eval_loss as for example in [NeMo](https://github.com/NVIDIA/NeMo/blob/33f5b9f98e8f48d37ca1980861fe630103598090/nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py#L271)

raphaelreinauer commented on gdeep/trainer/trainer.py:
```diff
@@ -24,6 +34,9 @@
 from gdeep.utility import _inner_refactor_scalars
 from gdeep.utility import DEVICE
+
+if TYPE_CHECKING:
```
Why is this needed?

hkirvesl replied:
avoid circular import

raphaelreinauer replied:
Since `Trainer` is not needed anymore in `regularizer.py`, you don't have circular imports anymore.

raphaelreinauer commented on gdeep/trainer/regularizer.py:
```diff
@@ -0,0 +1,434 @@
+    return dummy_loader
+
+
+def _unique_list(a_list: list) -> list:
```
This can be much simplified, see my previous comment.

hkirvesl replied:
yes, if hashable. Which is not the case here

raphaelreinauer commented on gdeep/trainer/regularizer.py:
```diff
@@ -0,0 +1,434 @@
+    return coordinates
+
+
+class Regularizer:
```
In my opinion, the usage of the abstract class `Regularizer` is complicating things more than it simplifies. Both `TopologicalRegularizer` and `TihonovRegularizer` have a very different interface. That's also the reason why the `__init__` method has as argument `**kwargs` and not some properly defined parameters.

The only thing that `TopologicalRegularizer` and `TihonovRegularizer` have in common is that they have a `regularization_penalty` method. Hence, I would build an interface for that. One option would be to define an abstract class only having an *abstract* `regularization_penalty` method:
```
from abc import ABC, abstract_method
class Regularizer(ABC):
    @abstract_method
    def regularization_penalty(self, model: nn.Module) -> Tenor:
        pass
class TopologicalRegularizerData(Regularizer):
    ...

class TihonovRegularizer(Regularizer):
    ...
```
 What I would prefer is to use a protocol instead, then you don't have the inherit from `Regularizer`:
```
from typing import Protocol
class Regularizer:
    def regularization_penalty(self, model: nn.Module) -> Tenor:
        pass

class TopologicalRegularizer:
    ...

class TihonovRegularizer:
    ...
```

hkirvesl replied:
done (the 2nd approach)

raphaelreinauer reviewed (changes requested):
I added some suggestions for the simplifying implementation.

hkirvesl commented:
Again, thanks for the comments and sorry to keep you waiting.

The idea on using protocols is spot on, that is exactly what we want here, and I've modified it accordingly.

I tried the idea on having 2 loss functions, but I think it gets quite complicated to get it work right with hyper parameter optimization. I think it is cleaner to have the penalty present whenever we are training, rather than specifying each time we call trainer which loss to use. 

The unique_list is indeed needed, and TYPE_CHECKING is to avoid circular imports

As a new features, I have also added support for the regularizer implementation for the benchmark. This logic tries to imitate the logic used for models and datasets, the end-user experience is the same. Details in /search/benchmark.py

hkirvesl commented on gdeep/trainer/trainer.py:
```diff
@@ -304,7 +320,23 @@ def closure() -> Tensor:
             batch_metric = self.training_metric(pred, y)
             metric_list.append(batch_metric)
-            loss = self.loss_fn(pred, y)
+            if self.regularize == True:
```
What is proposed here now morally does the same. The loss is also used in validation, so we don't want to overwrite it. With the 2 losses approach i think things will start to look quite ugly, especially in parameter optimization

raphaelreinauer commented on gdeep/trainer/regularizer.py:
```diff
@@ -0,0 +1,442 @@
+from gdeep.trainer import Trainer
```
That's not needed anymore.

hkirvesl replied:
right, done

raphaelreinauer commented on gdeep/trainer/trainer.py:
```diff
@@ -132,6 +147,7 @@ def forward(self, x):
     registered_hook: Optional[
         Callable[[int, Optimizer, ModelExtractor, Optional[SummaryWriter]], Any]
     ] = None
+    regularizer: Optional["Regularizer"] = None
```
```suggestion
    regularizer: Optional[Regularizer] = None
```

hkirvesl replied:
true, fixed now

raphaelreinauer commented on gdeep/trainer/trainer.py:
```diff
@@ -304,7 +326,13 @@ def closure() -> Tensor:
             batch_metric = self.training_metric(pred, y)
             metric_list.append(batch_metric)
-            loss = self.loss_fn(pred, y)
+            # loss = self.loss_fn(pred,y)
+            if self.regularizer is not None:
+                loss = self.loss_fn(pred, y)
+                penalty = self.regularizer.regularization_penalty(self.model)
+                loss = loss + penalty
+            else:
+                loss = self.loss_fn(pred, y)
             # Save to tensorboard
             try:
                 self.writer.add_scalar(  # type: ignore
```
```suggestion
            loss = self.loss_fn(pred,y)
            if self.regularizer is not None:
                penalty = self.regularizer.regularization_penalty(self.model)
                loss += penalty
           
            # Save to tensorboard
            try:
                self.writer.add_scalar(  # type: ignore
```

raphaelreinauer reviewed (changes requested):
It's looking much better now. Thanks for the changes.

hkirvesl commented:
Thanks for the comments, Raphael! Good suggestions took them all

matteocao reviewed (approved):
Thank you @hkirvesl for the great work! For me, this is approved.

Thank you @raphaelreinauer for the thorough review!! Feel free to share your final comments and/or approval.

Thank you again

raphaelreinauer commented:
LGTM! Thanks, @hkirvesl, for all the effort. The implementation looks much cleaner now.

raphaelreinauer commented:
Just approved your PR @hkirvesl .
